### PR TITLE
Add dedicated "minimal" mode for templates

### PR DESCRIPTION
**Pull request by cbentejac**
## Description

Following #1727, project files can be used as templates, with the .mg files containing the entire description of the graph. This implies that if there is a single node whose description is modified (parameters added, parameters removed, etc.), description conflicts will arise when opening a template containing such a node. For these conflicts to be resolved, the template would need to be saved again with the upgraded version of the node every single time there is a change, which defeats the purpose of having templates.

This PR adds a "minimal" mode to save and load templates. Instead of saving the entire graph description, the minimal mode only saves the modified input/output parameters. When that template is opened, conflicts are detected only on the saved parameters (with non-default values) and automatically solved. This makes templates robust to changes in node descriptions.

## Features list

- [x] Support saving only modified input/output parameters of a graph in an .mg file
- [x] Support reading an .mg file containing only modified input/output parameters
- [x] Add a "Save As Template" option that saves a graph in minimal mode
- [x] Update all the existing templates with minimal mode
- [x] Reload the list of available templates after saving a graph as a template 

## Implementation remarks

- Files written in minimal mode can be identified with the flag "template: True" in the header. If that flag is not detected in an .mg file, it is loaded as a regular file and not a template.
- .mg files written normally (with all the parameters saved) can still be opened as templates, but they will not be robust to any node description changes.
- test_multiviewPipeline now tests both "minimal" and "regular" modes: the first part of the test opens the default Photogrammetry template (which has been updated to minimal mode) while the second part of the test saves the graph obtained after reading the template into a normal .mg file (in which all the input/output parameters are written).

### CODE REVIEW

File: meshroom/core/attribute.py [hunk 1]
```diff
@@ -283,6 +283,8 @@ def defaultValue(self):
         return copy.copy(self.desc.value)
 
     def _isDefault(self):
+        if self.isLink:
+            return self.getLinkParam().value == self.defaultValue()
         return self._value == self.defaultValue()
```
Comment (fabiencastan) [inline meshroom/core/attribute.py:288]:
```suggestion
        return self.value == self.defaultValue()
```
Not fully sure if it's evaluated properly, else:
`self._get_value() == self.defaultValue()`

File: meshroom/core/graph.py [hunk 2]
```diff
@@ -1015,17 +1023,50 @@ def save(self, filepath=None, setupProjectFile=True):
...
+                Graph.IO.Keys.Graph: self.toDict()
+            }
 
         with open(path, 'w') as jsonFile:
             json.dump(data, jsonFile, indent=4)
 
         if path != self._filepath and setupProjectFile:
             self._setFilepath(path)
 
+    def getModifiedKeys(self):
```
Comment (fabiencastan) [inline meshroom/core/graph.py:1045]: getModifiedKeys(self) => getNonDefaultAttributes(self)

We are not returning only keys (myDict.keys(), myDict.values()).

File: meshroom/core/node.py [hunk 3]
```diff
@@ -1465,3 +1465,78 @@ def nodeFactory(nodeDict, name=None):
             node = node.upgrade()
 
     return node
+
+def templateNodeFactory(nodeDict, name=None):
```
Comment (fabiencastan) [inline meshroom/core/node.py:1469]: A bit too much duplication.
Maybe we can just add options to the nodeFactory function.
Using a functor for the filtering condition and an option to force nodes upgrade.

**Review by fabiencastan — CHANGES_REQUESTED** [review]
.

File: meshroom/core/attribute.py (hunk 1 above)
Comment (cbentejac) [reply meshroom/core/attribute.py]: Done

File: meshroom/core/graph.py (hunk 2 above)
Comment (cbentejac) [reply meshroom/core/graph.py]: Done

File: meshroom/core/node.py [hunk 4]
```diff
@@ -1463,5 +1464,7 @@ def nodeFactory(nodeDict, name=None):
         if not internalFolder and nodeDesc:
             logging.warning("No serialized output data: performing automatic upgrade on '{}'".format(name))
             node = node.upgrade()
+        if template:  # if the node comes from a template file and there is a conflict, it should be upgraded anyway
```
Comment (fabiencastan) [inline meshroom/core/node.py:1467]: Maybe `elif` to ensure that we are not calling the upgrade twice in any case?

Comment (cbentejac) [reply meshroom/core/node.py]: Done

**Review by fabiencastan — APPROVED** [review]

**Pull request merged**